### DISCUSSION TITLE
Light Sensor

### DISCUSSION THREAD

**Discussion by bearkillerPT** [expo/expo#18101] [Proposals and Ideas]
Hey there! I'm currently, happly, working in a expo project and using expo-sensors. The "boss" wants me to access the light sensor to detect ambient light and without taking a picture to detect it! I would like to collaborate but I'm new to bare RN and am still oblivious on how to start. Do you think this would be something easy to implement?

**Comment by barthap** [comment]
Hello!

I'm not sure how "new" you are to the native code at all, so a few things:
- I'm assuming that you know how to open a RN project in Xcode/Android Studio and browse files in it. Also, [contributing guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md) will be helpful
- As expo-sensors doesn't have a typical RN/expo module structure (the package has multiple native modules/services), it would be good if you learn it on a simple module - for example [expo-battery](https://github.com/expo/expo/tree/main/packages/expo-battery). Just read/follow its code on each platform and see how it works.
- The simplest platform to implement is probably Android - [look at this guide](https://developer.android.com/guide/topics/sensors/sensors_environment) and then for example at [BarometerModule.kt](https://github.com/expo/expo/blob/main/packages/expo-sensors/android/src/main/java/expo/modules/sensors/modules/BarometerModule.kt) and [BarometerService.kt](https://github.com/expo/expo/blob/main/packages/expo-sensors/android/src/main/java/expo/modules/sensors/services/BarometerService.kt). Most of the job is **probably** (I haven't dug deep into this) copy-pasting these files and replacing `TYPE_PRESSURE` with `TYPE_LIGHT` - but that's not all. Then you need to register your new module/service in SensorsPackage.
- Web should be simple too, although it is experimental and requires implementing asking for additional `ambient-light-sensor` permission. [See this page for details](https://developer.mozilla.org/en-US/docs/Web/API/AmbientLightSensor).
- The most tricky platform will probably be iOS. From what I briefly read, in the past there wasn't an official way to access this sensor (there was sth like `IOKit` but Apple rejected from App Store for using it). Some people used camera recording / AR kit for that ([discussion here](https://stackoverflow.com/questions/45440192/reading-ios-light-sensor-with-arkit)) but you said you don't want to use camera for that.
- Now there is a [iOS 14+ SensorKit framework](https://developer.apple.com/documentation/sensorkit) which has ambient light data, but has some stupid legal policy and requires tons of entitlements/provisioning/signing stuff 🤯  so we definitely do not want this in expo-sensors, but creating a 3rd party [expo-module](https://docs.expo.dev/modules/overview/) with [config plugin](https://docs.expo.dev/guides/config-plugins/) would be awesome, but this (the config plugin) is much more advanced stuff. For bare app you can still configure it manually though, without writing a config plugin.

**Reply by bearkillerPT** [reply]
Would it be okay for me to try to submit a pull request with the feature only working for android?

**Reply by barthap** [reply]
Sure, no problem :D You just need to mark in documentation, that it is an Android-only feature. Just ping me in the PR and I'll help you out 😉
